### PR TITLE
Balancer: Stable Pools in Baseline Solver.

### PR DESCRIPTION
**Pull request by bh2smith**
Putting the pieces together for Stable Pools as Baseline liquidity sources.

- [x] implement swap interface
- [x] deduplicate some of the code in swap interface
- [x] Still need to test `get_spot_price` - **moved to issue (https://github.com/gnosis/gp-v2-services/issues/1331)**
- [x] price estimation - **Reserved for a follow up PR.**

### Test Plan

Introduced new test for both new swap methods introduced. Namely

- `stable_get_amount_out`
- `stable_get_amount_in`

### CODE REVIEW

**Review by vkgnosis — APPROVED** [review]
Rust code looks good to me. Less confident about the balancer specifics but I don't see anything wrong either.

**Review by nlordell — COMMENTED** [review]
LGTM, just a missing TODO and a concern with rounding issues (where we use a different `balances` array than the SC does because we don't preserve order of the reserves).

File: shared/src/sources/balancer/swap.rs [hunk 1]
```diff
@@ -165,7 +158,124 @@ impl BaselineSolvable for WeightedPoolRef<'_> {
...
+            token_index_out,
+            out_reserves.upscale(out_amount)?,
+        )
+        .ok()?;
+        let amount_in_before_fee = in_reserves.downscale_up(in_amount).ok()?;
+        add_swap_fee_amount(amount_in_before_fee, self.swap_fee_percentage).ok()
+    }
+
+    /// Here we use the spot price defined in the white paper, assuming pools have equal weight.
+    /// https://balancer.fi/whitepaper.pdf#spot-price
```
[the remark this answers is not in the document]
Comment (bh2smith) [reply shared/src/sources/balancer/swap.rs:254]: Turns out, we will need to implement this differently

(cf https://github.com/gnosis/gp-v2-services/issues/1331)

File: shared/src/sources/balancer/swap.rs [hunk 2]
```diff
@@ -165,7 +158,116 @@ impl BaselineSolvable for WeightedPoolRef<'_> {
...
+    balances: Vec<Bfp>,
+}
+
+impl StablePoolRef<'_> {
+    fn construct_balances_and_token_indices(
+        &self,
+        in_token: &H160,
+        out_token: &H160,
+    ) -> Option<BalancesWithIndices> {
+        let mut balances = vec![];
```
[the remark this answers is not in the document]
Comment (bh2smith) [reply shared/src/sources/balancer/swap.rs:194]: https://github.com/gnosis/gp-v2-services/issues/1332

File: shared/src/sources/balancer/swap.rs [hunk 3]
```diff
@@ -326,4 +500,111 @@ mod tests {
...
+        let amount_in = 900_816_325_i128;
+        let amount_out = 900_000_000_000_000_000_000_u128.into();
+        let res_out = pool.get_amount_in(usdc, (amount_out, dai));
+        assert_eq!(res_out.unwrap(), amount_in.into());
+    }
+
+    #[test]
+    fn stable_spot_price() {
+        // TODO - write this test.
+    }
```
Comment (nlordell) [inline shared/src/sources/balancer/swap.rs:609]: ubernit: Can probable delete this empty test.

**Review by nlordell — APPROVED** [review]
LGTM

**Pull request merged**